File: server/src/db/schema/want.ts
import {
  pgTable,
  pgEnum,
  uuid,
  text,
  timestamp,
  primaryKey,
} from 'drizzle-orm/pg-core'

import { profile, topic } from './identity.ts'
import { responseValue } from './activity.ts'

// A loose, shared pre-activity: an intent + the friends you'd do it with, captured
// before there's a time/place. See specs/data-model.md (want) + specs/api/wants.md.
// Privacy is derived: no invitees ⇒ private; inviting a friend shares it with them.
export const wantKind = pgEnum('want_kind', ['one_shot', 'ongoing'])

export const want = pgTable('want', {
  id: uuid('id').primaryKey().defaultRandom(),
  ownerId: uuid('owner_id')
    .notNull()
    .references(() => profile.id, { onDelete: 'cascade' }),
  // Required: every want carries the noun-verb taxonomy so it's "ready to go".
  activityTypeId: uuid('activity_type_id')
    .notNull()
    .references(() => topic.id),
  title: text('title'),
  location: text('location'),
  notes: text('notes'),
  // one_shot archives once promoted; ongoing spawns activities repeatedly.
  kind: wantKind('kind').notNull().default('one_shot'),
  archivedAt: timestamp('archived_at', { withTimezone: true }),
  createdAt: timestamp('created_at', { withTimezone: true }).notNull().defaultNow(),
  updatedAt: timestamp('updated_at', { withTimezone: true }).notNull().defaultNow(),
})

// A friend the owner tagged onto a want, with that friend's soft response. Reuses
// the activity response vocabulary (in/interested/next_time); null = invited-not-yet-
// responded. ignoredAt set ⇒ the invitee ignored it (silent + recoverable; never
// surfaced to the owner — see the dismissal-is-silent principle).
export const wantInvite = pgTable(
  'want_invite',
  {
    wantId: uuid('want_id')
      .notNull()
      .references(() => want.id, { onDelete: 'cascade' }),
    profileId: uuid('profile_id')
      .notNull()
      .references(() => profile.id, { onDelete: 'cascade' }),
    response: responseValue('response'),
    ignoredAt: timestamp('ignored_at', { withTimezone: true }),
    createdAt: timestamp('created_at', { withTimezone: true }).notNull().defaultNow(),
  },
  (t) => [primaryKey({ columns: [t.wantId, t.profileId] })],
)

File: app/lib/models/want.dart
/// A loose, shared pre-activity (specs/api/wants.md, screens/wants.md): an intent
/// (topic + optional title/location/notes) plus the friends you'd do it with.
/// Tolerant reader — unknown fields ignored.
class Want {
  const Want({
    required this.id,
    this.ownerId,
    this.ownerName,
    this.activityTypeId,
    this.activityTypeLabel,
    this.title,
    this.location,
    this.notes,
    this.kind = 'one_shot',
    this.invitees = const [],
    this.yourResponse,
    this.archivedAt,
  });

  final String id;
  final String? ownerId;
  final String? ownerName;
  final String? activityTypeId;
  final String? activityTypeLabel;
  final String? title;
  final String? location;
  final String? notes;
  final String kind; // 'one_shot' | 'ongoing'
  final List<WantInvitee> invitees;
  final String?
  yourResponse; // set when you're an invitee: in|interested|next_time|null
  final DateTime? archivedAt;

  bool get isOngoing => kind == 'ongoing';
  bool get isArchived => archivedAt != null;

  factory Want.fromJson(Map<String, dynamic> json) {
    final owner = json['owner'] as Map<String, dynamic>?;
    final type = json['activity_type'] as Map<String, dynamic>?;
    return Want(
      id: json['id'] as String,
      ownerId: owner?['id'] as String?,
      ownerName: owner?['first_name'] as String?,
      activityTypeId: type?['id'] as String?,
      activityTypeLabel: type?['label'] as String?,
      title: json['title'] as String?,
      location: json['location'] as String?,
      notes: json['notes'] as String?,
      kind: json['kind'] as String? ?? 'one_shot',
      invitees: ((json['invitees'] as List<dynamic>?) ?? const [])
          .map((e) => WantInvitee.fromJson(e as Map<String, dynamic>))
          .toList(),
      yourResponse: json['your_response'] as String?,
      archivedAt: json['archived_at'] == null
          ? null
          : DateTime.tryParse(json['archived_at'] as String),
    );
  }
}

/// A friend invited to a want, with their soft response (null = not yet responded).
class WantInvitee {
  const WantInvitee({
    required this.profileId,
    this.name,
    this.photo,
    this.response,
  });

  final String profileId;
  final String? name;
  final String? photo;
  final String? response; // 'in' | 'interested' | 'next_time' | null

  factory WantInvitee.fromJson(Map<String, dynamic> json) {
    final p = json['profile'] as Map<String, dynamic>?;
    return WantInvitee(
      profileId: p?['id'] as String? ?? '',
      name: p?['first_name'] as String?,
      photo: p?['photo'] as String?,
      response: json['response'] as String?,
    );
  }
}
